components/TextControls/src/BaseTextAreas/MDCBaseTextArea.m

```diff
@@ -52,43 +94,460 @@ - (void)dealloc {
+                                      weakSelf.label.frame = weakSelf.labelFrame;
+                                    }
+                                  }];
+}
+
+- (BOOL)canLabelFloat {
+  return self.labelBehavior == MDCTextControlLabelBehaviorFloats;
+}
+
+- (MDCTextControlLabelState)determineCurrentLabelState {
+  return [self labelStateWithLabelText:self.label.text
+                                  text:self.textView.text
+                         canLabelFloat:self.canLabelFloat
+                             isEditing:self.textView.isFirstResponder];
+}
+
+- (MDCTextControlLabelState)labelStateWithLabelText:(NSString *)labelText
+                                               text:(NSString *)text
+                                      canLabelFloat:(BOOL)canLabelFloat
+                                          isEditing:(BOOL)isEditing {
+  BOOL hasLabelText = labelText.length > 0;
+  BOOL hasText = text.length > 0;
+  if (hasLabelText) {
+    if (canLabelFloat) {
+      if (isEditing) {
+        return MDCTextControlLabelStateFloating;
+      } else {
+        if (hasText) {
+          return MDCTextControlLabelStateFloating;
+        } else {
+          return MDCTextControlLabelStateNormal;
+        }
+      }
+    } else {
+      if (isEditing) {
+        return MDCTextControlLabelStateNone;
+      } else {
+        if (hasText) {
+          return MDCTextControlLabelStateNone;
+        } else {
+          return MDCTextControlLabelStateNormal;
+        }
+      }
+    }
+  } else {
+    return MDCTextControlLabelStateNone;
+  }
+}
+
 #pragma mark Custom Accessors
 
 - (UITextView *)textView {
-  return self.textAreaTextView;
+  return self.baseTextAreaTextView;
+}
+
+- (UILabel *)leadingAssistiveLabel {
+  return self.assistiveLabelView.leadingAssistiveLabel;
+}
+
+- (UILabel *)trailingAssistiveLabel {
+  return self.assistiveLabelView.trailingAssistiveLabel;
+}
+
+#pragma mark MDCTextControl Accessors
+
+- (void)setContainerStyle:(id<MDCTextControlStyle>)containerStyle {
+  id<MDCTextControlStyle> oldStyle = _containerStyle;
+  if (oldStyle) {
+    [oldStyle removeStyleFrom:self];
+  }
+  _containerStyle = containerStyle;
+  [_containerStyle applyStyleToTextControl:self animationDuration:self.animationDuration];
+}
+
+#pragma mark Fonts
+
+- (UIFont *)normalFont {
+  return self.baseTextAreaTextView.font ?: MDCTextControlDefaultUITextFieldFont();
+}
+
+- (UIFont *)floatingFont {
+  return [self.containerStyle floatingFontWithNormalFont:self.normalFont];
 }
 
 #pragma mark MDCBaseTextAreaTextViewDelegate
 
 - (void)textAreaTextView:(MDCBaseTextAreaTextView *)textView
-    willBecomeFirstResponder:(BOOL)willBecome {
-  [self setNeedsLayout];
+    willBecomeFirstResponder:(BOOL)didBecome {
+  if (textView == self.baseTextAreaTextView) {
+    [self setNeedsLayout];
+  }
 }
 
 - (void)textAreaTextView:(MDCBaseTextAreaTextView *)textView
-    willResignFirstResponder:(BOOL)willResign {
+    willResignFirstResponder:(BOOL)didResign {
```

Comment:
variable name is odd and doesn't match api

Reply:
Fixed! Thanks